drivers/display/ssd1673.c

```diff
@@ -225,118 +207,93 @@ static int ssd1673_write(const struct device *dev, const u16_t x,
 	}
 
-	if (x != 0 && y != 0) {
-		LOG_ERR("Unsupported origin");
-		return -1;
+	if ((y + desc->height) > EPD_PANEL_HEIGHT) {
+		LOG_ERR("Buffer out of bounds (height)");
+		return -EINVAL;
 	}
 
+	if ((x + desc->width) > EPD_PANEL_WIDTH) {
+		LOG_ERR("Buffer out of bounds (width)");
+		return -EINVAL;
+	}
 
-	ssd1673_busy_wait(driver);
-	memset(dummy_page, 0xff, sizeof(dummy_page));
+	if ((desc->height % EPD_PANEL_NUMOF_ROWS_PER_PAGE) != 0) {
+		LOG_ERR("Buffer height not multiple of %d",
+				EPD_PANEL_NUMOF_ROWS_PER_PAGE);
+		return -EINVAL;
+	}
+
+	if ((y % EPD_PANEL_NUMOF_ROWS_PER_PAGE) != 0) {
+		LOG_ERR("Y coordinate not multiple of %d",
+				EPD_PANEL_NUMOF_ROWS_PER_PAGE);
+		return -EINVAL;
+	}
 
 	switch (driver->scan_mode) {
 	case SSD1673_DATA_ENTRY_XIYDY:
-		if (ssd1673_set_ram_param(driver,
-					  SSD1673_PANEL_FIRST_PAGE,
-					  SSD1673_PANEL_LAST_PAGE + 1,
-					  SSD1673_PANEL_LAST_GATE,
-					  SSD1673_PANEL_FIRST_GATE)) {
-			return -1;
-		}
-
-		if (ssd1673_set_ram_ptr(driver,
-					SSD1673_PANEL_FIRST_PAGE,
-					SSD1673_PANEL_LAST_GATE)) {
-			return -1;
-		}
-
+		x_start = y / SSD1673_PIXELS_PER_BYTE;
+		x_end = (y + desc->height - 1) / SSD1673_PIXELS_PER_BYTE;
+		y_start = (x + desc->width - 1);
+		y_end = x;
 		break;
 
 	case SSD1673_DATA_ENTRY_XDYIY:
-		if (ssd1673_set_ram_param(driver,
-					  SSD1673_PANEL_LAST_PAGE + 1,
-					  SSD1673_PANEL_FIRST_PAGE,
-					  SSD1673_PANEL_FIRST_GATE,
-					  SSD1673_PANEL_LAST_GATE)) {
-			return -1;
-		}
-
-		if (ssd1673_set_ram_ptr(driver,
-					SSD1673_PANEL_LAST_PAGE + 1,
-					SSD1673_PANEL_FIRST_GATE)) {
-			return -1;
-		}
-
+		x_start = (EPD_PANEL_HEIGHT - y) / SSD1673_PIXELS_PER_BYTE;
+		x_end = (EPD_PANEL_HEIGHT - (y + desc->height - 1)) /
+			SSD1673_PIXELS_PER_BYTE;
+		y_start = x;
+		y_end = (x + desc->width - 1);
 		break;
 	default:
-		return -1;
+		return -EINVAL;
 	}
 
-	if (ssd1673_write_cmd(driver, SSD1673_CMD_ENTRY_MODE,
-			      &driver->scan_mode, sizeof(driver->scan_mode))) {
-		return -1;
-	}
+	ssd1673_busy_wait(driver);
 
-	gpio_pin_write(driver->dc, DT_SSD1673_DC_PIN, 0);
-	if (spi_write(driver->spi_dev, &driver->spi_config, &buf_set)) {
-		return -1;
+	err = ssd1673_write_cmd(driver, SSD1673_CMD_ENTRY_MODE,
+			&driver->scan_mode, sizeof(driver->scan_mode));
+	if (err < 0) {
+		return err;
 	}
 
-	gpio_pin_write(driver->dc, DT_SSD1673_DC_PIN, 1);
-	/* clear unusable page */
-	if (driver->scan_mode == SSD1673_DATA_ENTRY_XDYIY) {
-		sbuf.buf = dummy_page;
-		sbuf.len = sizeof(dummy_page);
-		if (spi_write(driver->spi_dev, &driver->spi_config, &buf_set)) {
-			return -1;
-		}
+	err = ssd1673_set_ram_param(driver, x_start, x_end, y_start, y_end);
+	if (err < 0) {
+		return err;
 	}
 
-	sbuf.buf = (u8_t *)buf;
-	sbuf.len = desc->buf_size;
-	if (spi_write(driver->spi_dev, &driver->spi_config, &buf_set)) {
-		return -1;
+	err = ssd1673_set_ram_ptr(driver, x_start, y_start);
+	if (err < 0) {
+		return err;
 	}
 
-	/* clear unusable page */
```

Comment:
That must stay, you can not really use these lines.

Reply:
There is no need to clear this lines on every write as it is done once on init